[Diagnostics] Generalize the "protocol type cannot conform" error to work for all types that cannot conform to protocols.

hborla opened the pull request:
Add a tailored diagnostic for missing conformance failures for (all) types that cannot conform to protocols.  Example:

```swift
protocol P {}

func genericFunc<T: P>(_ x: T) {}

genericFunc((1, "hello")) // error
```
 Previously, we diagnosed `argument type '(Int, String)' does not conform to expected type 'P'`. Now, we diagnose `type '(Int, String)' cannot conform to 'P'; only struct/enum/class types can conform to protocols`. Additionally, this change adds notes pointing to the requirement declaration, which we previously didn't show for existential types.

rdar://problem/44204160

hborla committed:
[Diagnostics] Generalize the "protocol type cannot conform" error to
work for all types that cannot conform to protocols.

hborla committed:
[Test] Add test cases for missing protocol conformance failures for
non-nominal types.

hborla commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -570,13 +570,26 @@ bool MissingConformanceFailure::diagnoseAsError() {
+  else if (nonConformingType->isExistentialType())
+    prohibitedTypeKind = "protocol";
+  else if (nonConformingType->is<AnyMetatypeType>())
+    prohibitedTypeKind = "metatype";
```
In my opinion we don't really need to spell this the kind of type here since they are all visually distinct e.g. metatype is going to have `.Type` at the end, protocol `.Protocol` etc.

hborla replied:
`.Type` and `.Protocol` are both metatypes, and the "protocol type" distinction here is for existential types. I agree that for functions, tuples, and metatypes, the type itself is enough to tell you what _kind_ of type it is, but I think it's useful for existential types because they look the same as struct/class/enum types. What do you think?

theblixguy replied:
How about “Non-nominal type %0 cannot conform to protocol %1”? Then you don’t need this extra logic.

hborla replied:
That was actually what I wrote originally, but "nominal" is a term only used in the compiler (it's not mentioned anywhere in TSPL) and many Swift programmers don't know what it means, especially beginners!

xedin replied:
I agree that it might make sense to have a special case for existentials something
`"{value of protocol |}%0type %1 cannot conform to %2; only struct/enum/class "
      "types can conform to protocols", (bool, Type, Type))`

theblixguy replied:
Yeah, that’s true. I remembered we have a similar diagnostic when extending a type like “Any” (non nominal type %0 cannot be extended) - perhaps it should be changed as well (separately, of course).

It would’ve been nice to have method that took a `TypeKind` and gave you back a string, similar to `DeclKind`/`DescriptiveDeclKind`, but I’m not sure if there’s a lot of existing code that has similar logic to the one you’ve got.

xedin replied:
I think it doesn't becomes when it comes to types most of the time (excluding existential types) it is self-evident what it is.

xedin commented on test/Constraints/function_builder_diags.swift:
```diff
@@ -166,7 +166,7 @@ struct Label<L> : P where L : P { // expected-note {{'L' declared as parameter t
 
 func test_51167632() -> some P {
-  AnyP(G { // expected-error {{static method 'buildBlock' requires that 'Label<Any>.Type' conform to 'P'}}
+  AnyP(G { // expected-error {{metatype type 'Label<Any>.Type' cannot conform to 'P'; only struct/enum/class types can conform to protocols}}
```
Loosing a note here is a regression, I think it's always useful to point exactly where the requirement is specified and what the substitution was if possible...

xedin commented on validation-test/compiler_crashers_2_fixed/0196-rdar48937223.swift:
```diff
@@ -2,10 +2,10 @@
 
 func test(str: String) {
-  fn(str) { arg in // expected-error {{global function 'fn(_:arg2:)' requires that '()' conform to 'P'}}
+  fn(str) { arg in // expected-error {{tuple type '()' cannot conform to 'P'; only struct/enum/class types can conform to protocols}}
```
In here as well, it would be great to preserve the note which points to declaration.

xwu commented:
> Previously, we diagnosed “argument type '(Int, String)' does not conform to expected type 'P'”. Now, we diagnose “tuple type '(Int, String)' cannot conform to 'P'; only struct/enum/class types can conform to protocols”.

I like the attempt to improve the diagnostic, but I’m not sure the revised wording is ideal:

“Argument type” is helpful here because it helps the reader figure out where the error specifically is that we’re talking about. It may not always be obvious that the type being referenced is that of the argument used, since the argument won’t always be a tuple literal. Consider a use site `f(x)`, where `x` is a tuple: the diagnostic would mention the generic constraints of `f` and the type of `x`, neither of which can be seen at the location where the error is displayed.

By contrast, consider the revised wording “tuple type ‘(Int, String)’”: it is clear to anyone with a basic knowledge of Swift that “(Int, String)” is a tuple type, so that wording is redundant.

Meanwhile, it is true that the tuple type “does not” conform to the protocol “P”, and it is equally true that it “cannot.”

However, the original diagnostic is phrased in terms of the specifics of why there is an error: the _argument to this specific function_ must be of a type that conforms to “P”—and this in turn implies the only way to call this function correctly: by supplying an argument of a type that does conform to ‘P’. It also implies a currently impossible fix when the argument is a tuple: conforming the type of the argument being used to the protocol. This is what we need to clarify.

Put another way, there are three pieces of information that need to be conveyed here to the user:

1. You have supplied an argument of the wrong type.
2. The argument is of the wrong type because there is a generic constraint requiring a protocol conformance.
3. There is no protocol conformance because tuple types cannot conform to any protocols.

The current diagnostic only accomplishes (1) and (2).

The proposed diagnostic here does not accomplish (1) because it does not phrase itself in terms of the specific argument. In fact, not only does it remove the word “argument,” it also phrases the error as a general rule about the language (“cannot”). It is a legitimate technique to teach users in error messages about what does or does not work in the language. But in this case it is phrased as a general rule with an oddly ambiguous specificity: now, the user is being taught that _this specific type_ cannot (as a rule) conform to _protocol “P”_. At least two implied fixes that a user might take away are false: another tuple type cannot conform to “P” any more than this one can, and this tuple type cannot conform to any other protocol any more than to “P”. Therefore, the proposed error message here correctly clarifies what it’s trying to say by then restating the general rule in a more accurate way. But then it does so by affirmatively mentioning structs, enums, and classes, none of which are relevant to this error in this place.

The end result is that a general rule is stated twice, accomplishing points (2) and (3) above more than once, while abandoning point (1). 

One wording that might work to teach users all three points is:
“Argument type ‘(Int, String)’ does not conform to expected protocol ‘P’; tuple types cannot conform to protocols.”

xedin commented:
@xwu Thank you for such a thoughtful comment! I agree that it would be best to get an additional variant of this diagnostic which is tailored for arguments as you suggest. That would actually be as easy as adjusting existing `cannot_convert_argument_value_protocol` diagnostic message because we already can distinguish that type of failure happening in argument position.

hborla commented:
@xwu again, thank you for the feedback! I agree that there is a lot of useful information that should be shown to the user, and this is where Swift's one-line diagnostic messages really fall short. FWIW, this is already what we show for existential types (which is arguably the most common form of this error), so we are already missing a lot of information in that case, and I'm attempting to remedy this by adding notes to this diagnostic. In the most recent version of Xcode, the error message for the example in the description actually says `Global function 'genericFunc' requires that '(Int, String)' conform to 'P`', along with a note pointing to the declaration of `genericFunc`. This is different than the current message produced on master, but it IS useful to know that the requirement comes from global function `genericFunc`, and this is information that the user should know. Additionally, I do think it's useful to state that only struct/enum/class types can conform to protocols because until the user changes the code so that they're sending in a value of a struct, enum, or class, they cannot fix the error.

As I mentioned, I'm currently adding notes to this diagnostic so that the user _does_ see more of this information (and I'm removing the redundancy of `tuple type`, `function type`, etc - see the thread above). IMO, ideally we would also show the original error message as a note in addition to the note pointing back to the declaration if applicable. So, the error message would say `type '(Int, String)' cannot conform to 'P'; only struct/enum/class types can conform to protocols` with notes stating _why_ it must conform, and _where_ the expected conformance is specified. I can also adjust the message to say "argument type" if applicable, but I'll have to figure out how to adjust this message when we have an existential type. We do still need to specify when a type is an existential type, because those types look just like any nominal type.

hborla committed:
[Diagnostics] Adjust wording of type_cannot_conform error message.

xwu commented:
How about "type '(Int, String)' cannot satisfy expected protocol constraint because it is not a struct, enum, or class"?

Again, I think for some users it can be misleading to say that it "cannot conform to 'P'", since `P` has nothing to do with what can or cannot happen here.

hborla committed:
[Diagnostics] Add notes for the type_cannot_conform error that point
to the declaration that requires protocol conformance.

hborla committed:
[Diagnostics] Adjust the description of existential types for the
type_cannot_conform diagnostic message.

hborla committed:
Merge remote-tracking branch 'upstream/master' into type-cannot-conform-diag

hborla committed:
[Test] Update Parse/confusables.swift with the "type cannot conform"
error message and note.

hborla committed:
[Diagnostics] In `MissingConformanceFailure`, if the non-conforming type
comes from an argument, use the argument expression as the anchor for
`diagnoseTypeCannotConform`.

hborla commented:
@swift-ci please smoke test
